spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala

```diff
@@ -0,0 +1,157 @@
+      table: RowLevelOperationTable,
+      assignments: Seq[Assignment],
+      cond: Option[Expression]): ReplaceData = {
+
+    // resolve all needed attrs (e.g. metadata attrs for grouping data on write)
+    val metadataAttrs = resolveRequiredMetadataAttrs(relation, table.operation)
+
+    // construct a read relation and include all required metadata columns
+    val readAttrs = dedupAttrs(relation.output ++ metadataAttrs)
+    val readRelation = relation.copy(table = table, output = readAttrs)
+
+    // build a plan with updated rows
+    val updateCond = cond.getOrElse(Literal.TrueLiteral)
+    val allRowsPlan = buildUpdateProjection(readRelation, assignments, updateCond)
+
+    // build a plan to replace read groups in the table
+    val writeRelation = relation.copy(table = table)
+    ReplaceData(writeRelation, allRowsPlan, relation)
+  }
+
+  // build a rewrite plan for sources that support replacing groups of data (e.g. files, partitions)
+  // if the condition contains a subquery
+  private def buildReplaceDataWithUnionPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      assignments: Seq[Assignment],
+      cond: Option[Expression]): ReplaceData = {
+
+    // resolve all needed attrs (e.g. metadata attrs for grouping data on write)
+    val metadataAttrs = resolveRequiredMetadataAttrs(relation, table.operation)
+
+    // construct a read relation and include all required metadata columns
+    // the same read relation will be used to read records that must be updated and be copied over
+    // DeduplicateRelations will take care of duplicated attr IDs
+    val readAttrs = dedupAttrs(relation.output ++ metadataAttrs)
+    val readRelation = relation.copy(table = table, output = readAttrs)
+
+    // build a plan for records that match the cond and should be updated
+    val updateCond = cond.getOrElse(Literal.TrueLiteral)
+    val matchedRowsPlan = Filter(updateCond, readRelation)
+    val updatedRowsPlan = buildUpdateProjection(matchedRowsPlan, assignments)
+
+    // build a plan for records that did not match the cond but had to be copied over
+    val remainingRowFilter = Not(EqualNullSafe(updateCond, Literal.TrueLiteral))
+    val remainingRowsPlan = Filter(remainingRowFilter, readRelation)
+
+    // new state is a union of updated and copied over records
+    val allRowsPlan = Union(updatedRowsPlan, remainingRowsPlan)
+
+    // build a plan to replace read groups in the table
+    val writeRelation = relation.copy(table = table)
+    ReplaceData(writeRelation, allRowsPlan, relation)
+  }
+
+  // this method assumes the assignments have been already aligned before
+  // the condition passed to this method may be different from the UPDATE condition
+  private def buildUpdateProjection(
+      plan: LogicalPlan,
+      assignments: Seq[Assignment],
+      cond: Expression = Literal.TrueLiteral): LogicalPlan = {
+
+    // TODO: avoid executing the condition for each column
```

Comment:
It wouldn't be too hard to add a projection with one more column and then use the column value to determine whether to replace values.

Reply:
Unfortunately, the optimizer would squash the projections and replace the attribute with its expression. We had this limitation in the old implementation too.

Reply:
If this is what we'd end up with anyway, then let's hold off. I see that this is what we do in 3.1 anyway so no need to change it.

Reply:
I think we would need a custom node to handle it. I'll look into it after the release.